src/DataMapper/DataMapper.php

```diff
@@ -637,11 +637,10 @@ private function calculateChanges(array &$changes, array $newData, array $oldDat
 
         foreach ($oldData as $k => $v) { // data that used to exist, but now doesn't
             if (!isset($newData[$k])) { // removed field
-                if (is_integer($k)) {
+                if (is_integer($k) && array_values($oldData) === $oldData) { // is integer and not associative array
```

Comment:
O ajuste não pode gerar `breaking change`?
Talvez seja válido criar uma versão 3.2 para esse ajuste

Reply:
Good question.
I think will not have breaking change, because this only avoid associative indexes with numbers, like `[1500857923]`, that actually is a product code.

The main goal of this workaround is that the MongoDB can't remove fields with number indexes like `items[1]`, instead of remove this index `1`, the MongoDB let this index with `null` value.
So the ending of the update process let the document like this:
```
"shipments": [
            {
              ...
              "items": [
                ...
                null
             ]
...
```

It is an expected behavior that isn't happening.
That's why I think is not a `breaking change`.

Reply:
I believe it is more performant to do this "computation" if the array is a list outside the foreach, so it won't be run for each key in the array.

Also, I suggest using the package [symfony/polyfill-php81](https://github.com/symfony/polyfill-php81), it provides the function [array_is_list](https://www.php.net/manual/ en/function.array-is-list.php) available in PHP 8.1 that does the same validation.

Indirectly we have already used several of these polyfills, including the one from PHP 8.0 installed by other dependencies of composer.json, as seen in composer.lock.